### DISCUSSION TITLE
Brush fixes & features

### DISCUSSION THREAD

**Discussion by ShaMan123** [fabricjs/fabric.js#7134] [Ideas]
I was thinking of 2 things I have in mind for brushes:
### Behavior with canvas overlays
Should canvas overlays render on top of the brush while drawing or not? I think they should. Current behavior is inconsistent from the developer's point of view. While drawing, the brush paints above the overlay (was this intentional?). Once drawing is done the path is created and added to canvas so it renders underneath the overlay as all objects should and do. I don't like this.
`EraserBrush` has this fix embedded in it because of it's logic.
It is relatively simple. 
Approaches:
1. If there's an overlay we hide it when rendering the main context and draw it with the brush on the top context. 
2. We can fix rendering order in canvas, making overlay render last

Another thought regarding this and consistency. What about active objects? How do they interact with overlays?

### Clipping
I have added in my projects a feature for PencilBrush to draw in confined areas, for example letting the brush draw only on top of a certain object or area. I am willing to PR and saw  interest regarding this #6955.
I'm not sure how to expose this functionality in a proper and meaningful way. I use a method `bindtoObjects` which makes the brush draw only on top of these objects. But it is too specific. 
Perhaps the right approach would be to expose a `clipPath` property on brush. This will enable all the above and much more. Plus it is consistent with the rest of fabric in this term.

@asturur @melchiar What do you think?

**Comment by asturur** [comment]
Yes those points deserve some discussion.
We should start defining what is the scope of the overlay itself.
The whole upperCanvas is just a speed optimisation and we should keep this in mind.
In general, consistency and robustness is better than speed optimisations.

So what the overlay is supposed to do?

For me the overlay is a set once and forget item where i want to define a work area/ a print area.
The object on the top of the stack, set with evented: false, is basically an overlay.
It can also be a frame, or a static/dynamic ui of some sort

The overlay is something you want to paint but don't have your users mess with it.

The overlay could be moved on the upperCanvas when we move controls there.

Also an opinion: object controls below the overlay do not make sense, i m not sure if you can still paint them in that way with the option `controlsAboveOverlay: false`

**Reply by ShaMan123** [reply]
I agree.
Why did you consolidate all special rendering to one canvas? why not a canvas for each breaking upper canvas to 4? 
1. controls
2. active selection
3. brush 
4. overlay

Then something like `controlsAboveOverlay: false` can be handled by ordering the canvas stack

**Reply by ShaMan123** [reply]
I think this makes sense with clip path advanced usage
